workspaces/vscode-python/Dockerfile

```diff
@@ -0,0 +1,109 @@
+        --install-extension oderwat.indent-rainbow \
+		# visualize csv files in vscode window
+		--install-extension mechatroner.rainbow-csv \
+		# linting for markdown files
+		--install-extension DavidAnson.vscode-markdownlint \
+		# format tables in markdown files
+        --install-extension fcrespo82.markdown-table-formatter \
+	) && \
+	rm -rf /home/coder/.local/share/code-server/CachedExtensionVSIXs
+
+# Prepare the entrypoints. The entrypoint.sh part is based on what code-server
+# originally had, so it may need to be updated when code-server is updated. We
+# avoid running fixuid if the user is root (for local testing in PL). We also
+# disable auth as PL handles it.
+USER root
+RUN rm -f /run/fixuid.ran && \
+	echo "" > /usr/bin/entrypoint.sh && chmod 0755 /usr/bin/entrypoint.sh && \
+	/usr/bin/env echo -e '#!/usr/bin/env bash \n' \
+	'set -eu \n' \
+	'# We do this first to ensure sudo works below when renaming the user. \n' \
+	'# Otherwise the current container UID may not exist in the passwd database. \n' \
+	'[ $(id -u) -ne 0 ] && eval "$(fixuid -q)" \n' \
+	' \n' \
+	'if [ "${DOCKER_USER-}" ]; then \n' \
+	'  USER="$DOCKER_USER" \n' \
+	'  if [ "$DOCKER_USER" != "$(whoami)" ]; then \n' \
+	'    echo "$DOCKER_USER ALL=(ALL) NOPASSWD:ALL" | sudo tee -a /etc/sudoers.d/nopasswd > /dev/null \n' \
+	'    # Unfortunately we cannot change $HOME as we cannot move any bind mounts \n' \
+	'    # nor can we bind mount $HOME into a new home as that requires a privileged container. \n' \
+	'    sudo usermod --login "$DOCKER_USER" coder \n' \
+	'    sudo groupmod -n "$DOCKER_USER" coder \n' \
+	' \n' \
+	'    sudo sed -i "/coder/d" /etc/sudoers.d/nopasswd \n' \
+	'  fi \n' \
+	'else \n' \
+	'  sudo rm /etc/sudoers.d/nopasswd \n' \	
+	'fi \n' \
+	'\n' \
+	'dumb-init /usr/bin/code-server --auth none "$@" \n' >> /usr/bin/entrypoint.sh
+RUN rm /requirements.txt
+
+EXPOSE 8080
+
+# TODO: Maybe customize the preferences to set the theme to auto-adjust to the
+# browser, turn off more telemetry, etc.
+
+# Please read the note here carefully to avoid wiping out what you installed
+# in ~/.local:
+
+# EDITOR_FOCUS_DIR should be set to the directory you want the editor to start
+# up in. This is not necessarily the same as the "home" setting in the
+# question's info.json file. The "home" setting determines the mount point for
+# the persistent cloud storage, which will hide any contents your image
+# originally had at the same path. You might want to set both EDITOR_FOCUS_DIR
+# and "home" to a deeper directory than /home/coder if you want to keep the
+# default home contents from the workspace image (~/.local, etc.). For
+# example, using /home/coder/question will copy the question's workspace
+# folder contents into an empty mount at /home/coder/question and save it for
+# the student, while always reusing the initial contents of /home/coder that
+# you prepared in the image. (However, if students try to customize their
```

Comment:
This talks about _maybe_ setting `EDITOR_FOCUS_DIR` to something inside `/home/coder`, but below it's already set to `/home/coder/project`. Is this intentional? This feels like it might cause projects where the default of `home` is `/home/coder`; starter files would be copied to e.g. `/home/coder/fibonacci.py`, but the editor would focus on `/home/coder/project` (which would not in fact exist on the mounted-in volume). Or does this new version now require that one sets `"home": "/home/coder/project"`?

Reply:
Thinking about the "preserving student settings" issue, I think it's not worth setting the question home to `/home/coder` just for that, because the settings will get reset when moving between questions or question instances anyway, and the name-hiding issue of intentional image contents in `/home/coder` might cause problems later on. To preserve student settings, it would be cool if PL had a feature where they could provide a json for that in their site-wide user settings, and have it auto-injected into a workspace. But there would be other compatibility/security issues with that, maybe.

Reply:
Also, I made a commit below that does use `/home/coder/workspace` instead of `/home/coder/project`.

Reply:
> To preserve student settings, it would be cool if PL had a feature where they could provide a json for that in their site-wide user settings, and have it auto-injected into a workspace. But there would be other compatibility/security issues with that, maybe.

I've thought about this, a la the way GitHub Codespaces supports a `dotfiles` repository. The huge obvious problem for this is that this would provide a means to smuggle information into (and possible out of) testing facilities.

I think `/home/coder/workspace` is a reasonable default.